_posts/2023-02-22-user-defined-operators.md

```diff
@@ -0,0 +1,484 @@
+
+function uf(int16 mantissa, uint8 exponent) pure suffix returns (UFixed16x2) { ... }
+```
+```solidity
+UFixed16x2 circumference = 2 uf * 3.14 uf * radius;
+```
+
+While the application of a suffix will involve a function call, eventually, with the help of
+compile-time evaluation, literal suffixes are going to become as efficient as native literals.
+
+#### More Operators
+Many operators have been excluded from the initial implementation.
+Some of them will very likely be user-definable at some point.
+
+1. `<<`, `>>` and `**` were excluded due to their non-uniform arguments.
+    For built-in types, the right-hand side is always an unsigned type.
+1.  `!` was excluded because we have to decide whether it should return the same type as the argument
+    of just `bool`.
+    We are leaning towards the latter, but it needs more consideration.
+1. `++` and `--` are very convenient, and we want to have them eventually.
+    They just need to be handled in a special way, and there is more than way to do so.
+1. Composite assignment operators, such as `+=`, will likely be provided automatically
+    in the future as long as the corresponding arithmetic/bitwise operator is available.
+    Allowing a direct implementation is currently not impossible anyway, because there are no references
+    to value types in the language.
+1. We are considering adding a C++-like
+    ["spaceship operator"](https://en.wikipedia.org/wiki/Three-way_comparison#High-level_languages),
+    i.e. `<=>`.
+    When defined, the compiler could use it to automatically generate all the other comparison operators.
+
+On the other hand, it is unlikely that we will allow redefining some other operators:
+
+1. `&&` and `||` are not necessary for most UDVTs because user-defined comparison operators
+    must return `bool`, for which the built-in boolean operators are sufficient.
+    The additional complication is their short-circuiting behavior that may or may not be
+    appropriate, depending on the semantics of a particular UDVT.
+1. `=` is already provided by the compiler for all types and it is very unlikely that we will ever allow redefining it.
+
+#### Operators on Reference Types
+While UDVTs were the primary target, not all types can reasonably fit into a single slot.
+The way to solve this problem is to allow operators taking and returning structs.
+The main thing holding us back are the planned changes to reference types, in particular the fact
+that [data location](https://docs.soliditylang.org/en/latest/types.html#data-location) is going to
+become a part of the type itself.
+ To avoid having to introduce breaking changes later, we decided to defer support for structs until
+these changes are behind us.
+
+We want to eventually support operators on other reference types, but it is likely that this usage will
+require naming them, using some kind of yet-to-be-introduced user-defined reference type feature similar
+to UDVTs, or at least defining a type alias.
```

Comment:
This is a bit much conjecturing... it's equally possible that we allow user-defined operators on structs themselves, so I'd just skip this part.

Reply:
Removed.